### PR TITLE
removing our image and forcing setting of an ami

### PR DESCRIPTION
**Pull request by chrislovecnm**
Tweaking the script to force the setting of an ami.  We should know which ami we are developing against anyways :) Called it $IMAGE because --image.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1031)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
softball .. have at it :)

**Comment by brandoncole** [comment]
LGTM!

File: hack/dev-build.sh [hunk 1]
```diff
@@ -45,6 +45,7 @@ KOPS_DIRECTORY="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 [ -z "$KOPS_STATE_STORE" ] && echo "Need to set KOPS_STATE_STORE" && exit 1;
 [ -z "$CLUSTER_NAME" ] && echo "Need to set CLUSTER_NAME" && exit 1;
 [ -z "$NODEUP_BUCKET" ] && echo "Need to set NODEUP_BUCKET" && exit 1;
+[ -z "$IMAGE" ] && echo "Need to set IMAGE or use the image list https://github.com/kubernetes/kops/blob/master/channels/stable" && exit 1;
```
Comment (kris-nova) [inline hack/dev-build.sh]: Something tells me that this script will be ran from a tool like Jenkins, and the exit code might matter.

Do we want to keep all the missing variable exit codes == `1`?

Comment (kris-nova) [inline hack/dev-build.sh]: Also do we want to default this? Or at least have a way to easily access a common image? Enumerate them?

```
IMAGE=2 sh dev-build.sh
```

```
if [ "$IMAGE" -eq "2" ]; then
   IMAGE="common.image.here"
fi
```

Comment (chrislovecnm) [inline hack/dev-build.sh]: Awesome fill fix.

Comment (chrislovecnm) [inline hack/dev-build.sh]: Will fix ;)

**Comment by chrislovecnm** [comment]
@k8s-ci-robot test this

File: hack/dev-build.sh (hunk 1 above)
Comment (chrislovecnm) [inline hack/dev-build.sh]: I re-read this and I am confused.  I would like to force the user to specify an image, be the user is a developer.  Thoughts?

**Comment by chrislovecnm** [comment]
@k8s-bot aws e2e test this

**Comment by chrislovecnm** [comment]
@kris-nova will 
`NODEUP_URL="https://s3-us-west-1.amazonaws.com/${NODEUP_BUCKET}/kops/${GIT_VER}/linux/amd64/nodeup"`

Always work or do we need to specify the aws zone url?

**Comment by chrislovecnm** [comment]
We could do something nutty like this and curl the yaml from the stable file and parse it for the value:

```bash
function parse_yaml {
   local prefix=$2
   local s='[[:space:]]*' w='[a-zA-Z0-9_]*' fs=$(echo @|tr @ '\034')
   sed -ne "s|^\($s\):|\1|" \
        -e "s|^\($s\)\($w\)$s:$s[\"']\(.*\)[\"']$s\$|\1$fs\2$fs\3|p" \
        -e "s|^\($s\)\($w\)$s:$s\(.*\)$s\$|\1$fs\2$fs\3|p"  $1 |
   awk -F$fs '{
      indent = length($1)/2;
      vname[indent] = $2;
      for (i in vname) {if (i > indent) {delete vname[i]}}
      if (length($3) > 0) {
         vn=""; for (i=0; i<indent; i++) {vn=(vn)(vname[i])("_")}
         printf("%s%s%s=\"%s\"\n", "'$prefix'",vn, $2, $3);
      }
   }'
}
```

File: hack/dev-build.sh (hunk 1 above)
Comment (kris-nova) [inline hack/dev-build.sh]: If we would like to force the user to specify an image, than ignore my comments :) I thought having enumerated common images would be a win

**Comment by kris-nova** [comment]
@chrislovecnm 

# Nodeup URL
From the [S3 documentation](http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro)

> Amazon S3 supports both virtual-hosted–style and path-style URLs to access a bucket.
> 
> In a virtual-hosted–style URL, the bucket name is part of the domain name in the URL. For example:  
>     http://bucket.s3.amazonaws.com
>     http://bucket.s3-aws-region.amazonaws.com.
> In a virtual-hosted–style URL, you can use either of these endpoints. If you make a request to the http://bucket.s3.amazonaws.com endpoint, the DNS has sufficient information to route your request directly to the region where your bucket resides.

So let's be super slick about this.. and point `NODEUP_URL`like:

```
NODEUP_URL="https://${NODEUP_BUCKET}.s3.amazonaws.com/kops/${GIT_VER}/linux/amd64/nodeup"
```

Example: https://kops-devel.s3.amazonaws.com/kops/git-60dce75/linux/amd64/nodeup

# The parse YAML function

I don't understand what we are parsing, and why we need it. More clarity please :) 

But brownie points for a good lucking `bash` function.. Haven't seen `sed` and `awk` like that in a while!

**Comment by chrislovecnm** [comment]
The yaml is coming from https://github.com/kubernetes/kops/blob/master/channels/stable ~ where we could get the latest stable channel ami.  Overkill me thinks.  I appropriated the bash from stackoverflow, so I will not take the credit 😀

**Comment by kris-nova** [comment]
If you think the YAML parsing takes less time to implement than we would spend manually overriding it in the next 3 revisions - I would say go for it - otherwise, probably a sign it's over kill.

Do you see the YAML parsing saving our $youKnowWhat in the future ever?

**Comment by chrislovecnm** [comment]
nah ... yaml is overkill ..  Working on testing.

**Comment by chrislovecnm** [comment]
@kris-nova take a final look at merge please.

**Comment by chrislovecnm** [comment]
@k8s-bot aws e2e test this

File: hack/dev-build.sh [hunk 2]
```diff
@@ -23,29 +23,38 @@
...
 # https://aws.amazon.com/cli/
-# and make sure you `aws configure` 
+# and make sure you `aws configure`
 #
 # Example usage
 #
 # KOPS_STATE_STORE="s3://my-dev-s3-state \
 # CLUSTER_NAME="fullcluster.name.mydomain.io" \
 # NODEUP_BUCKET="s3-devel-bucket-name-store-nodeup" \
+# IMAGE="kope.io/k8s-1.4-debian-jessie-amd64-hvm-ebs-2016-10-21" \
```
Comment (justinsb) [inline hack/dev-build.sh]: FYI there is 2016-12-05 IIRC

**Comment by justinsb** [comment]
LGTM

**Pull request merged**